Doesn't work in Spring Boot jar

Spring BootJar has complex layout and special spring ClassLoader to make it work
SmallRye Config doesn't work in Spring Application packed as spring bootJar.

I have found, the problem is in your class
`io.smallrye.common.classloader.ClassPathUtils`

Full explanation and fix:
https://github.com/smallrye/smallrye-config/issues/1101

PS: more advanced research and fix are welcome 🙏

@radcortez @dmlloyd Sorry to bother you 🙏
But the situation is critical. SmallRye Config doesn't work in a typical Spring Boot application (bundled as spring boot jar). 
Could it be fixed in the near future?

```
git clone https://github.com/magicprinc/SmallRyeConfig-SpringBoot.git
cd SmallRyeConfig-SpringBoot
./gradlew runBootJar
```
⇒
```
> Task :runBootJar FAILED
Exception in thread "main" java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:118)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:108)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:58)
        at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:65)
Caused by: java.lang.ExceptionInInitializerError
        at fink.demo.smallryeconfigspringboot.ExampleApplication.main(ExampleApplication.java:13)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        ... 5 more
Caused by: java.io.UncheckedIOException: Failed to read C:\TEMP\test\SmallRyeConfig-SpringBoot\build\libs\SmallRyeConfig-SpringBoot-1.0.jar!\BOOT-INF\classes
        at io.smallrye.common.classloader.ClassPathUtils.processAsPath(ClassPathUtils.java:146)
        at io.smallrye.common.classloader.ClassPathUtils.consumeAsPath(ClassPathUtils.java:102)
        at io.smallrye.common.classloader.ClassPathUtils.consumeAsPaths(ClassPathUtils.java:86)
        at io.smallrye.config.AbstractLocationConfigSourceLoader.tryClassPath(AbstractLocationConfigSourceLoader.java:141)
        at io.smallrye.config.AbstractLocationConfigSourceLoader.loadConfigSources(AbstractLocationConfigSourceLoader.java:104)
        at io.smallrye.config.AbstractLocationConfigSourceLoader.loadConfigSources(AbstractLocationConfigSourceLoader.java:87)
        at io.smallrye.config.PropertiesConfigSourceProvider.<init>(PropertiesConfigSourceProvider.java:37)
        at fink.config.spring.SmallRyeConfigAutoConf.<clinit>(SmallRyeConfigAutoConf.java:75)
        ... 7 more
Caused by: java.nio.file.NoSuchFileException: C:\TEMP\test\SmallRyeConfig-SpringBoot\build\libs\SmallRyeConfig-SpringBoot-1.0.jar!\BOOT-INF\classes
        at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.<init>(ZipFileSystem.java:166)
        at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.getZipFileSystem(ZipFileSystemProvider.java:125)
        at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.newFileSystem(ZipFileSystemProvider.java:120)
        at java.base/java.nio.file.FileSystems.newFileSystem(FileSystems.java:528)
        at java.base/java.nio.file.FileSystems.newFileSystem(FileSystems.java:400)
        at io.smallrye.common.classloader.ClassPathUtils.processAsPath(ClassPathUtils.java:139)
        ... 14 more

FAILURE: Build failed with an exception.
```

--- a/classloader/src/main/java/io/smallrye/common/classloader/ClassPathUtils.java
+++ b/classloader/src/main/java/io/smallrye/common/classloader/ClassPathUtils.java
@@ -8,12 +8,12 @@ import java.net.URISyntaxException;
 import java.net.URL;
 import java.net.URLConnection;
 import java.nio.file.FileSystem;
-import java.nio.file.FileSystems;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.nio.file.spi.FileSystemProvider;
 import java.util.Enumeration;
+import java.util.Map;
 import java.util.function.Consumer;
 import java.util.function.Function;
 
@@ -105,6 +105,19 @@ public class ClassPathUtils {
         });
     }
 
+    private static final FileSystemProvider JAR_PROVIDER;
+
+    static {
+        final ClassLoader ccl = Thread.currentThread().getContextClassLoader();
+        try {
+            Thread.currentThread().setContextClassLoader(ClassLoader.getSystemClassLoader());
+            JAR_PROVIDER = FileSystemProvider.installedProviders().stream().filter(p -> p.getScheme().equals("jar")).findFirst()
+                    .orElseThrow();
+        } finally {
+            Thread.currentThread().setContextClassLoader(ccl);
+        }
+    }
+
     /**
      * Attempts to represent a resource as a local file system path to be processed by a function.
      * If a resource appears to be an actual file or a directory, it is simply passed to the function as-is.
@@ -118,32 +131,25 @@ public class ClassPathUtils {
      */
     public static <R> R processAsPath(URL url, Function<Path, R> function) {
         if (JAR.equals(url.getProtocol())) {
-            final ClassLoader ccl = Thread.currentThread().getContextClassLoader();
-            try {
-                // We are loading "installed" FS providers that are loaded from the system classloader anyway
-                // To avoid potential ClassCastExceptions we are setting the context classloader to the system one
-                Thread.currentThread().setContextClassLoader(ClassLoader.getSystemClassLoader());
-                FileSystemProvider.installedProviders();
-            } finally {
-                Thread.currentThread().setContextClassLoader(ccl);
-            }
-
             final String file = url.getFile();
-            final int exclam = file.lastIndexOf('!');
-            final Path jar;
+            final int exclam = file.indexOf('!');
             try {
-                jar = toLocalPath(exclam >= 0 ? new URL(file.substring(0, exclam)) : url);
+                URL fileUrl;
+                String subPath;
+                if (exclam == -1) {
+                    // assume the first element is a JAR file, not a plain file, since it was a `jar:` URL
+                    fileUrl = new URL(file);
+                    subPath = "/";
+                } else {
+                    fileUrl = new URL(file.substring(0, exclam));
+                    subPath = file.substring(exclam + 1);
+                }
+                if (!fileUrl.getProtocol().equals("file")) {
+                    throw new IllegalArgumentException("Sub-URL of JAR URL is expected to have a scheme of `file`");
+                }
+                return processAsJarPath(toLocalPath(fileUrl), subPath, function);
             } catch (MalformedURLException e) {
                 throw new RuntimeException("Failed to create a URL for '" + file.substring(0, exclam) + "'", e);
-            }
-            try (FileSystem jarFs = FileSystems.newFileSystem(jar, (ClassLoader) null)) {
-                Path localPath = jarFs.getPath("/");
-                if (exclam >= 0) {
-                    localPath = localPath.resolve(file.substring(exclam + 1));
-                }
-                return function.apply(localPath);
-            } catch (IOException e) {
-                throw new UncheckedIOException("Failed to read " + jar, e);
             }
         }
 
@@ -152,6 +158,35 @@ public class ClassPathUtils {
         }
 
         throw new IllegalArgumentException("Unexpected protocol " + url.getProtocol() + " for URL " + url);
+    }
+
+    private static <R> R processAsJarPath(Path jarPath, String path, Function<Path, R> function) {
+        try (FileSystem jarFs = JAR_PROVIDER.newFileSystem(jarPath, Map.of())) {
+            Path localPath = jarFs.getPath("/");
+            int start = 0;
+            for (;;) {
+                int idx = path.indexOf('!', start);
+                if (idx == -1) {
+                    return function.apply(localPath.resolve(path));
+                } else {
+                    // could be nested JAR?
+                    Path subPath = localPath.resolve(path.substring(0, idx));
+                    if (Files.isDirectory(subPath)) {
+                        // no, it's a plain directory and the `!` is superfluous
+                        localPath = subPath;
+                        start = idx + 1;
+                        if (start + 1 < path.length() && path.charAt(start + 1) == '/') {
+                            start++;
+                        }
+                    } else {
+                        // yes, it's a nested JAR file
+                        return processAsJarPath(subPath, path.substring(idx + 1), function);
+                    }
+                }
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException("Failed to read " + jarPath, e);
+        }
     }
 
     /**
@@ -204,7 +239,7 @@ public class ClassPathUtils {
 
     /**
      * Translates a URL to local file system path.
-     * In case the the URL couldn't be translated to a file system path,
+     * In case the URL couldn't be translated to a file system path,
      * an instance of {@link IllegalArgumentException} will be thrown.
      *
      * @param url URL

